internal/scheduler/scheduler.go

```diff
@@ -0,0 +1,264 @@
+package scheduler
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/boundary/internal/scheduler/job"
+	"github.com/hashicorp/go-hclog"
+	ua "go.uber.org/atomic"
+)
+
+type jobRepoFactory func() (*job.Repository, error)
+
+type runningJob struct {
+	runId     string
+	cancelCtx context.CancelFunc
+}
+
+// Scheduler is used to register and run recurring jobs on the controller.
+type Scheduler struct {
+	serverId  string
+	jobRepoFn jobRepoFactory
+	logger    hclog.Logger
+
+	registeredJobs *sync.Map
+	runningJobs    map[JobId]runningJob
```

Comment:
Any reason not to use a sync.Map here? I know below that there is a spot that locks the mutex and looks for existing value in the map but that can be handled with LoadOrStore on a sync.Map.

Reply:
Hmm, I could.  I had some more complex locking at first which I removed after some tweaks which means I could probably do this move as well...

Reply:
Generally speaking the mutex+lock is a good approach if you either have a bad use case for sync.Map (as they outline in the docs) or if you specifically want to stop execution until you get the lock. I believe in your cases you don't need to stop execution if it's already there (or not), just take different paths. If I'm correct then I do think it would simplify things a bit to use it. It's not complex now, but anytime you can remove dealing with a separate lock is usually for the better :-)